packages/react/src/components/NumberInput/NumberInput.js

```diff
@@ -199,8 +218,13 @@ class NumberInput extends Component {
           value,
         },
         () => {
-          this.props.onClick(evt, direction);
-          this.props.onChange(evt, direction);
+          if (useControlledStateWithValue) {
```

Comment:
We're using the flag here to change the second argument of onClick and onChange, rather than it's intended function of changing the controlled/uncontrolled behavior of components.

I'm okay with using it as a signal of this breaking change, but we need to be careful to control the scope of what our flags do. I see the return type change here as a separate, related change. But we just need to document it very well along with the flag's actual purpose.